src/raftstore/store/fsm/batch.rs

```diff
@@ -0,0 +1,586 @@
+            sender: channel::Sender<FsmTypes<N, C>>,
+        }
+
+        impl<N, C> Clone for $name<N, C> {
+            #[inline]
+            fn clone(&self) -> $name<N, C> {
+                $name {
+                    sender: self.sender.clone(),
+                }
+            }
+        }
+    };
+}
+
+impl_sched!(NormalScheduler);
+impl_sched!(ControlScheduler);
+
+impl<N: Fsm, C> FsmScheduler for NormalScheduler<N, C> {
+    type Fsm = N;
+
+    #[inline]
+    fn schedule(&self, fsm: Box<N>) {
+        match self.sender.send(FsmTypes::Normal(fsm)) {
+            Ok(()) => return,
+            // TODO: use debug instead.
+            Err(SendError(FsmTypes::Normal(fsm))) => warn!("failed to schedule fsm {:p}", fsm),
+            _ => unreachable!(),
+        }
+    }
+
+    fn shutdown(&self) {
+        // TODO: close it explicitly once it's supported.
+        // Magic number, actually any number greater than poll pool size works.
+        for _ in 0..100 {
+            let _ = self.sender.send(FsmTypes::Empty);
+        }
+    }
+}
+
+impl<N, C: Fsm> FsmScheduler for ControlScheduler<N, C> {
+    type Fsm = C;
+
+    #[inline]
+    fn schedule(&self, fsm: Box<C>) {
+        match self.sender.send(FsmTypes::Control(fsm)) {
+            Ok(()) => return,
+            Err(SendError(FsmTypes::Control(fsm))) => warn!("failed to schedule fsm {:p}", fsm),
+            _ => unreachable!(),
+        }
+    }
+
+    fn shutdown(&self) {
+        for _ in 0..100 {
+            let _ = self.sender.send(FsmTypes::Empty);
+        }
+    }
+}
+
+/// A basic struct for a round of polling.
+pub struct Batch<N, C> {
+    normals: Vec<Box<N>>,
+    control: Option<Box<C>>,
+}
+
+impl<N: Fsm, C: Fsm> Batch<N, C> {
+    /// Create a a batch with given batch size.
+    pub fn with_capacity(cap: usize) -> Batch<N, C> {
+        Batch {
+            normals: Vec::with_capacity(cap),
+            control: None,
+        }
+    }
+
+    fn push(&mut self, fsm: FsmTypes<N, C>) -> bool {
+        match fsm {
+            FsmTypes::Normal(n) => self.normals.push(n),
+            FsmTypes::Control(c) => {
+                assert!(self.control.is_none());
+                self.control = Some(c);
+            }
+            FsmTypes::Empty => return false,
+        }
+        true
+    }
+
+    fn is_empty(&self) -> bool {
+        self.normals.is_empty() && self.control.is_none()
+    }
+
+    fn clear(&mut self) {
+        self.normals.clear();
+        self.control.take();
+    }
+
+    /// Put back the FSM located at index.
+    ///
+    /// Only when channel length is larger than `checked_len` will trigger
+    /// further notification. This function may fail if channel length is
+    /// larger than the given value before FSM is released.
+    pub fn release(&mut self, index: usize, checked_len: usize) -> bool {
+        let mut fsm = self.normals.swap_remove(index);
+        let mailbox = fsm.take_mailbox().unwrap();
+        mailbox.release(fsm);
```

Comment:
Does it mean if any new mesages sent to the mailbox, the fsm won't get notified?